### PR TITLE
Add InfiniTime Vision document

### PR DESCRIPTION
**Pull request by Riksu9000**
Add the initial InfiniTIme Vision document that has been written in collaboration with the core developers. This document will help steer efforts towards a common goal by having the goals of the project written down and visible to everyone.

Since the last draft I added some context to the top of the document and using standard protocols as a core principle.

### CODE REVIEW

File: doc/InfiniTimeVision.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:1]: Let's put the `Vision` part as <h1> at the top of the document and make the other headers level 2 <h2>

```suggestion
# InfiniTime Vision

The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.
```

File: doc/InfiniTimeVision.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+# InfiniTime
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:3]: make level 2 header

```suggestion
## InfiniTime
```

File: doc/InfiniTimeVision.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+InfiniTime is a community built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.
+
+InfiniTime is not to be used for medical or other health tracking purposes.
+
+# Core Principles
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:9]: make level 2 header

```suggestion
## Core Principles
```

File: doc/InfiniTimeVision.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+- KISS
+- Reliability
+- Battery efficiency
+- Easy and simple navigation
+- Behaviour should be predictable and easy to understand
+- Prefer solid default experience over customization
+- Personalization is achieved through custom watchfaces. More options may be available through a companion app.
+- Use standard protocols and methods
+
+# Long term vision
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:20]: make level 2 header

```suggestion
## Long term vision
```

File: doc/InfiniTimeVision.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+# InfiniTime
+
+InfiniTime is a community built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.
+
+InfiniTime is not to be used for medical or other health tracking purposes.
+
+# Core Principles
+
+- KISS
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:11]: just to be sure everyone knows what we mean
```suggestion
- KISS (Keep It Short and Simple)
```

File: doc/InfiniTimeVision.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+- Reliability
+- Battery efficiency
+- Easy and simple navigation
+- Behaviour should be predictable and easy to understand
+- Prefer solid default experience over customization
+- Personalization is achieved through custom watchfaces. More options may be available through a companion app.
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:17]: for markdown files it is a good custom to have one line per sentence. Then diffs are easier to read and one word change will only change one line, instead of possible a whole paragraph

```suggestion
- Personalization is achieved through custom watchfaces.
  More options may be available through a companion app.
```

File: doc/InfiniTimeVision.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+# Long term vision
+
+The perfect version of InfiniTime would include:
+
+- Capability to upload apps and watchfaces from a phone
+- Only minimal features. Users would add the features they want.
+- Ports to other devices
+- Translations
+- Great documentation
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:28]: user documentation or developer documentation or both?

**Review by NeroBurner — CHANGES_REQUESTED** [review]

**Comment by devnoname120** [comment]
Are you open to feedback from non-core contributors?

**Comment by Riksu9000** [comment]
> Are you open to feedback from non-core contributors?

You're absolutely free to comment on anything, just note that during this PR we'll likely only be fixing possible flaws and not change the ideas.

File: doc/InfiniTimeVision.md (hunk 7 above)
Comment (NeroBurner) [reply doc/InfiniTimeVision.md:32]: I think yes, and we'll keep the code readable ourselfs ;)

File: doc/InfiniTimeVision.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+More options may be available through a companion app.
+- Use standard protocols and methods
+
+## Long term vision
+
+The perfect version of InfiniTime would include:
+
+- Capability to upload apps and watchfaces from a phone
+- Only minimal features.
+Users would add the features they want.
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:29]: It seems to render fine without spaces, but in the text it is easer to read if indented
```suggestion
  Users would add the features they want.
```

File: doc/InfiniTimeVision.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+- Personalization is achieved through custom watchfaces.
+More options may be available through a companion app.
+- Use standard protocols and methods
+
+## Long term vision
+
+The perfect version of InfiniTime would include:
+
+- Capability to upload apps and watchfaces from a phone
+- Only minimal features.
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:28]:
```suggestion
- Only minimal feature set in the flashed firmware.
```

File: doc/InfiniTimeVision.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
+# InfiniTime Vision
+
+The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.
+
+## InfiniTime
+
+InfiniTime is a community built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.
+
+InfiniTime is not to be used for medical or other health tracking purposes.
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:9]: We don't intend InfiniTime to be used as medical device, because we can't (and don't want) to have to meet such a high standard and responsibility. But everyone can do what he/she wants with InfiniTime.

Or is the "not to be used" part some legal speak?
```suggestion
InfiniTime is not intended to be used for medical or other health tracking purposes.
```

**Review by NeroBurner — CHANGES_REQUESTED** [review]

File: doc/InfiniTimeVision.md (hunk 10 above)
Comment (Riksu9000) [reply doc/InfiniTimeVision.md:9]: Of course everyone's free to do whatever regardless of what we say, but we want no part in that. I'd rather keep the statement strong and clear.

File: doc/InfiniTimeVision.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
+# InfiniTime Vision
+
+The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.
+
+## InfiniTime
+
+InfiniTime is a community built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.
```
Comment (FintasticMan) [inline doc/InfiniTimeVision.md:7]: Very minor grammatical issue, it should be community-built rather than community built.

File: doc/InfiniTimeVision.md (hunk 10 above)
Comment (NeroBurner) [reply doc/InfiniTimeVision.md:9]: Good point!

File: doc/InfiniTimeVision.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+- Prefer solid default experience over customization
+- Personalization is achieved through custom watchfaces.
+  More options may be available through a companion app.
+- Use standard protocols and methods
+
+## Long term vision
+
+The perfect version of InfiniTime would include:
+
+- Capability to upload apps and watchfaces from a phone
```
Comment (NeroBurner) [inline doc/InfiniTimeVision.md:27]: Non-Blocking: Not just phone, but PC as well

```suggestion
- Capability to upload apps and watchfaces from a phone or computer
```

**Review by NeroBurner — APPROVED** [review]

File: doc/InfiniTimeVision.md (hunk 12 above)
Comment (FintasticMan) [reply doc/InfiniTimeVision.md:27]: Perhaps it should be "using a companion app".

**Comment by devnoname120** [comment]
> You're absolutely free to comment on anything, just note that during this PR we'll likely only be fixing possible flaws and not change the ideas.

Thanks I'll try to give feedback within one week.

File: doc/InfiniTimeVision.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+- Only a minimal feature set in the flashed firmware.
```
Comment (devnoname120) [inline doc/InfiniTimeVision.md:28]: Doesn't this contradict the objective to `prefer solid default experience over customization`?
Imho we should remove non-essential stuff like pong, and move stuff like chimes to standalone apps. But useful features (find my phone, weather, calendar, etc.) should be available out of the box imho.

File: doc/InfiniTimeVision.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+## Core Principles
+
+- KISS (Keep It Short and Simple)
+- Reliability
+- Battery efficiency
+- Easy and simple navigation
+- Behaviour should be predictable and easy to understand
+- Prefer solid default experience over customization
+- Personalization is achieved through custom watchfaces.
```
Comment (devnoname120) [inline doc/InfiniTimeVision.md:19]: Currently watchfaces are really hard to implement. Designers can't implement them because most of them can't program, and developers are predominantly terrible at UI/UX.

Once InfiniTime has external storage and this kind of stuff I'll probs work on a specification to make the creation of watchfaces easy & flexible. I'll harness my experience as a product engineer, and my close collaboration with designers in my previous work.

File: doc/InfiniTimeVision.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+## InfiniTime
+
+InfiniTime is a community-built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.
+
+InfiniTime is not to be used for medical or other health tracking purposes.
+
+## Core Principles
+
+- KISS (Keep It Short and Simple)
```
Comment (devnoname120) [inline doc/InfiniTimeVision.md:13]: Imho this is vague. What about something like `write simple code, expand later if needed`. Since it's a mono-codebase it's useless to build supposedly ‶future-proof″ abstractions right away. We can refactor code once we have a clearer view about what kind of abstraction would be a great fit.

File: doc/InfiniTimeVision.md (hunk 14 above)
Comment (devnoname120) [reply doc/InfiniTimeVision.md:19]: > Personalization is achieved through custom watchfaces.

Also, I believe that it contradicts the statement `capability to upload apps`.

File: doc/InfiniTimeVision.md [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+- Reliability
+- Battery efficiency
+- Easy and simple navigation
+- Behaviour should be predictable and easy to understand
+- Prefer solid default experience over customization
+- Personalization is achieved through custom watchfaces.
+  More options may be available through a companion app.
+- Use standard protocols and methods
```
Comment (devnoname120) [inline doc/InfiniTimeVision.md:21]: Could we be more specific? For example BLE services with a link to a summary of the available ones. Maybe also explain the rational that justifies the burden of not just rolling out our custom easy-to-implement/debug solutions?

File: doc/InfiniTimeVision.md [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+## Long term vision
+
+The perfect version of InfiniTime would include:
+
+- Capability to upload apps and watchfaces from a phone
+- Only a minimal feature set in the flashed firmware.
+  Users would add the features they want.
+- Ports to other devices
+- Translations
```
Comment (devnoname120) [inline doc/InfiniTimeVision.md:31]: **Pitfall**: localizing apps is incredibly hard to do, and comes with a lot of unexpected challenges. PineTime requires users to be tech-savy — so they can overwhelmingly understand basic English. Is support for other languages worth it before a commercial pre-installed noob-friendly watch gets released?

Comment (Riksu9000) [reply doc/InfiniTimeVision.md:31]: There isn't a plan currently to implement translations, but it's certainly something that would be great to have someday, hence I think long term vision is the correct place for this.

File: doc/InfiniTimeVision.md (hunk 13 above)
Comment (Riksu9000) [reply doc/InfiniTimeVision.md:28]: It depends on the interpretation of minimal in this case. We intend to keep basic apps like for example stopwatch, timer and alarm. We'll need to consider which apps should be kept when the time comes, but that'll be a long time from now.

File: doc/InfiniTimeVision.md (hunk 14 above)
Comment (Riksu9000) [reply doc/InfiniTimeVision.md:19]: > Currently watchfaces are really hard to implement. Designers can't implement them because most of them can't program, and developers are predominantly terrible at UI/UX.

Makes sense. It would be great to get some collaboration happening between designers and developers.

> Also, I believe that it contradicts the statement capability to upload apps.

Hmm, how so?

File: doc/InfiniTimeVision.md (hunk 15 above)
Comment (Avamander) [reply doc/InfiniTimeVision.md:13]: It's intentionally so. Also "Future-proof" abstractions are rarely short and simple.

Comment (devnoname120) [reply doc/InfiniTimeVision.md:13]: @Avamander this is literally what I said.

File: doc/InfiniTimeVision.md (hunk 13 above)
Comment (ItzSwirlz) [reply doc/InfiniTimeVision.md:28]: I think he idea of 'minimal' is like nothing too crazy-for example no big bowling app game but something like Snake would be acceptable

File: doc/InfiniTimeVision.md (hunk 16 above)
Comment (JF002) [reply doc/InfiniTimeVision.md:21]: One of the goals of implementing standard and widely used protocols is to avoid re-inventing the wheel. Why spend time and efforts designing a protocol when we can just implement an existing one that fits our needs? Designing a protocol like DFU or file transfer protocol is not that easy. By using a well established one that is already used for similar use-case, we ensure that this protocol will probably work fine for us too.

Also, it makes the integration with other devices and apps a lot more easier. If a companion app already implements such a protocol, adding support for the PineTime will be virtually free. Why should we require companion apps developers to implement yet another protocol to set the time or to fetch heart rate values when they have already implement them multiple times?

In InfiniTime, we implement standard protocols from the BLE spec (GATT services : https://github.com/oesmith/gatt-xml).
For the BLE FS API, we implemented the BLE File Transfer Protocol from Adafruit (https://github.com/adafruit/Adafruit_CircuitPython_BLE_File_Transfer).
On the other hand, we couldn't find anything standard for the motion service, so we designed a custom one (https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/doc/MotionService.md)

We choose the protocol on a case-by-case basis, but designing and implementing a new one from scratch should always be considered as a last resort.

**Review by JF002 — APPROVED** [review]

File: doc/InfiniTimeVision.md (hunk 13 above)
Comment (Danimations) [reply doc/InfiniTimeVision.md:28]: Bring on snake... if this is a serious prospect, I'd love to produce art for it.

File: doc/InfiniTimeVision.md [hunk 18]
```diff
@@ -0,0 +1,32 @@
+# InfiniTime Vision
+
+The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.
+
+## InfiniTime
+
+InfiniTime is a community-built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.
```
Comment (Riksu9000) [inline doc/InfiniTimeVision.md:7]: @Danimations has suggested to rephrase this as:
```suggestion
InfiniTime is a community-built firmware for the PineTime smartwatch. It offers freedom and privacy advantages unavailable to users of other proprietary wearable technology.
```

Perhaps saying it's only for the PineTime is too specific however, as there will be ports in the future.

Comment (FintasticMan) [reply doc/InfiniTimeVision.md:7]: I personally think that what is already there is mostly good, but that specifying that the other wearable technology it mentions is the proprietary kind would improve it.

Comment (Danimations) [reply doc/InfiniTimeVision.md:7]: Given your feedback, perhaps this is appropriate:

"InfiniTime is a community-built smartwatch firmware. It offers freedom and privacy advantages unavailable to users of proprietary wearable technology."

File: doc/InfiniTimeVision.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (JF002) [reply doc/InfiniTimeVision.md:13]: I like this one!

File: doc/InfiniTimeVision.md (hunk 12 above)
Comment (Riksu9000) [reply doc/InfiniTimeVision.md:28]: How about this?
```suggestion
- Capability to sideload apps and watchfaces
```

File: doc/InfiniTimeVision.md (hunk 14 above)
Comment (devnoname120) [reply doc/InfiniTimeVision.md:20]: > Makes sense. It would be great to get some collaboration happening between designers and developers.

Great! I'll work on mockups before implementing it so as to make sure we are on the same page. :)

> Hmm, how so?

I interpret this statement as exhaustive about how we can customize the OS. What do you think? It does not explicitly state “only” but many users and contributors won't pick the subtlety of the wording.

Do we plan to add other customizations than watchfaces and apps?

If there are not too many, I'd suggest making this list exhaustive.

Comment (Riksu9000) [reply doc/InfiniTimeVision.md:20]: Ah I see what you mean. Here personalization means aesthetic, non-functional options, like changing fonts, certain colors etc. We've decided to limit aesthetic customization mainly to the watchface to keep things simple and try to provide a great user experience by default.

Many projects end up making a lot of things configurable, which quickly becomes overwhelming for users. We try to avoid adding too many options if possible, but not to the point of detriment.
